### PR TITLE
Pushing OperatorOverloading

### PR DESCRIPTION
**Pull request by tavisca-mdixit**
@amarpalsapure  Operator Overloading with Exception handling

@amarpalsapure  Please review my code ,, which is updated on the basis of comments received

### CODE REVIEW

File: 01 OperatorOverloading/OperatorOverlaoding.Money/Money.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace OperatorOverloadingMoney
```
Comment (tavisca-mdixit) [inline 01 OperatorOverloading/OperatorOverlaoding.Money/Money.cs]: Root Namespace should be same.

File: 01 OperatorOverloading/OperatorOverlaoding.Money/Money.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    {
+    public class Money
+    {
+
+        public double Amount { get; set; }
+        public string Currency { get; set; }
+
+        public static Money operator+(Money moneyOne,Money moneyTwo)  {
+            Money moneyThree = new Money();
```
Comment (tavisca-mdixit) [inline 01 OperatorOverloading/OperatorOverlaoding.Money/Money.cs]: Why construct object here?

File: 01 OperatorOverloading/OperatorOverlaoding.Money/Money.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+          
+                if (moneyOne.Currency.Equals(moneyTwo.Currency) )
```
Comment (tavisca-mdixit) [inline 01 OperatorOverloading/OperatorOverlaoding.Money/Money.cs]: What will happen if the either of parameter is null?

Comment (tavisca-mdixit) [inline 01 OperatorOverloading/OperatorOverlaoding.Money/Money.cs]: What will happen if `Currency` is not set?

Comment (tavisca-mdixit) [inline 01 OperatorOverloading/OperatorOverlaoding.Money/Money.cs]: What will happen if first currency is USD and second is usd?

File: 01 OperatorOverloading/OperatorOverlaoding.Money/Money.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+                {
+                    
+                    moneyThree.Amount= moneyOne.Amount + moneyTwo.Amount;
+                    if (moneyOne.Amount>(double.MaxValue-moneyTwo.Amount) || moneyThree.Amount<0)
+                    { throw new System.ArgumentException(); }
+                    moneyThree.Currency = moneyOne.Currency;
+                    return moneyThree;
+                }
+                else{
+                throw new System.Exception();}
```
Comment (tavisca-mdixit) [inline 01 OperatorOverloading/OperatorOverlaoding.Money/Money.cs]: Why there is no message in the exception? Ideally you should throw a custom exception.

File: 01 OperatorOverloading/OperatorOverlaoding.Money/Money.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        public double Amount { get; set; }
+        public string Currency { get; set; }
+
+        public static Money operator+(Money moneyOne,Money moneyTwo)  {
+            Money moneyThree = new Money();
+          
+                if (moneyOne.Currency.Equals(moneyTwo.Currency) )
+                {
+                    
+                    moneyThree.Amount= moneyOne.Amount + moneyTwo.Amount;
```
Comment (tavisca-mdixit) [inline 01 OperatorOverloading/OperatorOverlaoding.Money/Money.cs]: Why code is not formatted?

File: 01 OperatorOverloading/OperatorOverlaoding.Money/Money.cs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+                    if (moneyOne.Amount>(double.MaxValue-moneyTwo.Amount) || moneyThree.Amount<0)
```
Comment (tavisca-mdixit) [inline 01 OperatorOverloading/OperatorOverlaoding.Money/Money.cs]: Write a comment.

Comment (tavisca-mdixit) [inline 01 OperatorOverloading/OperatorOverlaoding.Money/Money.cs]: You can use `checked` here

File: 01 OperatorOverloading/OperatorOverloading.Host/Host.cs [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+using System.Threading.Tasks;
+using OperatorOverloadingMoney;
+
+namespace OperatorOverloading.Host
+{
+    class Host
+    {
+        static void Main(string[] args)
+        {
+            Money Money1 = new Money();
```
Comment (tavisca-mdixit) [inline 01 OperatorOverloading/OperatorOverloading.Host/Host.cs]: Naming Convention: local variables

File: 01 OperatorOverloading/OperatorOverloading.Host/Host.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+            Money Money2 = new Money();
+            Money Money3 = new Money();
+
+     
+
+            try
+            {
+                Console.WriteLine("Enter the Curreny type  USD/INR and the amount ");
+                Money1.Currency = Console.ReadLine();
+                Money1.Amount = Convert.ToDouble(Console.ReadLine());
```
Comment (tavisca-mdixit) [inline 01 OperatorOverloading/OperatorOverloading.Host/Host.cs]: Try to use`double.TryParse`

File: 01 OperatorOverloading/OperatorOverloading.Host/Host.cs [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    
+
+                Console.WriteLine("Enter the Curreny type  USD/INR/YEN and the amount ");
+                Money2.Currency = Console.ReadLine();
+                Money2.Amount = Convert.ToDouble(Console.ReadLine());
+                if (Money1.Amount < 0 || Money2.Amount < 0)
+                { throw new System.Exception();}
+                Money3 = Money2 + Money1;
+            }
+            catch(Exception E){
```
Comment (tavisca-mdixit) [inline 01 OperatorOverloading/OperatorOverloading.Host/Host.cs]: `E` should be in camel case

**Comment by tavisca-mdixit** [comment]
@amarpalsapure Please review my code ,, which is updated on the basis of comments received

File: 01 OperatorOverloading/OperatorOverlaoding.Money/Money.cs [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace OperatorOverloading
+{
+    public class Money
+    {
+
+        public double amount { get; set; }
```
Comment (tavisca-mdixit) [inline 01 OperatorOverloading/OperatorOverlaoding.Money/Money.cs]: Public property should be in pascal case

**Comment by tavisca-mdixit** [comment]
@amarpalsapure Code improved after the 2nd review,, you can review the code

File: 01 OperatorOverloading/OperatorOverlaoding.Money/Money.cs [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        public string Currency
+        {
+            get
+            {
+                return _currency;
+            }
+            private set
+            {
+                //Checking for Empty/Null Strings
+                if (string.IsNullOrEmpty(value) == true)
```
Comment (apalsapure) [inline 01 OperatorOverloading/OperatorOverlaoding.Money/Money.cs]: Use `IsNullOrWhiteSpace`

File: 01 OperatorOverloading/OperatorOverloading.Host/Host.cs [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+            Money moneyTwo;
+            Money moneyThree;
+
+            try
+            {
+
+
+                Console.WriteLine("Enter the Curreny type  USD/INR and the amount ");
+                currency = Console.ReadLine();
+                double.TryParse(Console.ReadLine(), out amount);
```
Comment (apalsapure) [inline 01 OperatorOverloading/OperatorOverloading.Host/Host.cs]: What if user types `abcd`? Your program will still execute with amount as zero. Is that correct?

File: 01 OperatorOverloading/OperatorOverloading.Host/Host.cs [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+                //Calling the Parameterized constructor of Money Class
+                moneyOne = new Money(amount,currency);
```
Comment (apalsapure) [inline 01 OperatorOverloading/OperatorOverloading.Host/Host.cs]: Format your code.

File: 01 OperatorOverloading/OperatorOverloading.Host/Host.cs [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+                Console.WriteLine("Enter the Curreny type  USD/INR/YEN and the amount ");
+                currency = Console.ReadLine();
+               
+
+
+                //Calling the Parameterized constructor of Money Class
+                moneyTwo = new Money(amount,currency);
+                
+                //Overloading the Plus operator           
+                moneyThree = moneyTwo + moneyOne;
```
Comment (apalsapure) [inline 01 OperatorOverloading/OperatorOverloading.Host/Host.cs]: Can this be written `var moneyThree = moneyTwo + moneyOne;`

File: 01 OperatorOverloading/OperatorOverloading.Host/Host.cs (hunk 12 above)
Comment (apalsapure) [inline 01 OperatorOverloading/OperatorOverloading.Host/Host.cs]: That's why a constructor which accepts `100 USD` helps the consumer. By doing so all the validation is done by the `Money` class.

**Comment by tavisca-mdixit** [comment]
@amarpalsapure  I have added a second constructor ,You can review the code.

**Comment by tavisca-mdixit** [comment]
@amarpalsapure  You can review and psuh into master

File: 01 OperatorOverloading/OperatorOverlaoding.Money/Money.cs [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+
+        public string Currency
+        {
+            get
+            {
+                return _currency;
+            }
+            private set
+            {
+                if ((value.Length == 3) == false)
```
Comment (apalsapure) [inline 01 OperatorOverloading/OperatorOverlaoding.Money/Money.cs]: run your code for `new Money(100, null)`

**Comment by tavisca-mdixit** [comment]
@amarpalsapure Corrected the code,based on the review.

**Comment by tavisca-mdixit** [comment]
@apalsapure Corrected the code, based upon the review.

**Pull request merged**